typing/typecore.ml

```diff
@@ -1941,17 +1941,64 @@ let check_univars env expans kind exp ty_expected vars =
   raise (Error (exp.exp_loc, env,
                 Less_general(kind, [Unification_trace.diff ty ty_expected])))
 
-(* Check that a type is not a function *)
-let check_application_result env statement exp =
-  let loc = exp.exp_loc in
-  match (expand_head env exp.exp_type).desc with
-  | Tarrow _ ->
-      Location.prerr_warning exp.exp_loc Warnings.Partial_application
-  | Tvar _ -> ()
-  | Tconstr (p, _, _) when Path.same p Predef.path_unit -> ()
-  | _ ->
-      if statement then
-        Location.prerr_warning loc Warnings.Statement_type
+let check_partial_application statement exp =
+  let rec f delay =
+    let ty = (expand_head exp.exp_env exp.exp_type).desc in
+    let check_statement () =
+      match ty with
+      | Tconstr (p, _, _)  when Path.same p Predef.path_unit ->
+          ()
+      | _ ->
+          if statement then
+            let rec loop {exp_loc; exp_desc; _} =
+              match exp_desc with
+              | Texp_let (_, _, e)
+              | Texp_sequence (_, e)
+              | Texp_try (e, _)
+              | Texp_ifthenelse (_, e, _)
+              | Texp_match (_, {c_rhs = e; _} :: _, _) ->
+                  loop e
+              | _ ->
+                  Location.prerr_warning exp_loc Warnings.Statement_type
+            in
+            loop exp
+    in
+    match ty, exp.exp_desc with
+    | Tarrow _, _ ->
+        let rec check {exp_desc; exp_loc; exp_extra; _} =
+          if List.exists (function
+              | (Texp_constraint _, _, _) -> true
+              | _ -> false) exp_extra then check_statement ()
+          else begin
+            match exp_desc with
+            | Texp_ident _ | Texp_constant _ | Texp_tuple _
+            | Texp_construct _ | Texp_variant _ | Texp_record _
+            | Texp_field _ | Texp_setfield _ | Texp_array _
+            | Texp_while _ | Texp_for _ | Texp_new _ | Texp_instvar _
+            | Texp_setinstvar _ | Texp_override _ | Texp_letmodule _
+            | Texp_letexception _ | Texp_assert _ | Texp_lazy _ | Texp_object _
+            | Texp_pack _ | Texp_unreachable | Texp_extension_constructor _
+            | Texp_ifthenelse (_, _, None) | Texp_function _ | Texp_send _ ->
+                check_statement ()
```

Comment:
If I understand correctly, the above are supposed to be cases which cannot be function applications, and as a result there is no need to warn for partial application.
However, the `let exception` and  `let module` have the type of their body, which may be a function application.
Similarly, `Texp_send` denotes a method call, which is itself a function application, and should raise a warning.

Reply:
I have added the `let exception` and `let module` cases, and fixed the handling of `Texp_send` and `Texp_new`, thanks.

The case of `Texp_field` was discussed [above](https://github.com/ocaml/ocaml/pull/1974#discussion_r231193986) and it was suggested that it should not raise an exception, which seems reasonably to me by analogy with the case of identifiers.